internal/dag/gatewayapi_processor.go

```diff
@@ -233,28 +233,86 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
 	return strings.ToLower(certificateRef.Kind) == "secret" && strings.ToLower(certificateRef.Group) == "core"
 }
 
-func (p *GatewayAPIProcessor) computeHosts(hostnames []gatewayapi_v1alpha1.Hostname) ([]string, []error) {
-	// Determine the hosts on the route, if no hosts
-	// are defined, then set to "*".
-	var hosts []string
+// computeHosts validates the hostnames for a HTTPRoute as well as validating
+// that the hostname on the HTTPRoute matches what is optionally defined on the
+// listener.hostname.
+func (p *GatewayAPIProcessor) computeHosts(hostnames []gatewayapi_v1alpha1.Hostname, listenerHostname *gatewayapi_v1alpha1.Hostname) (map[string]struct{}, []error) {
+
+	hosts := make(map[string]struct{})
 	var errors []error
-	if len(hostnames) == 0 {
-		hosts = append(hosts, "*")
+
+	// Determine the hosts on the hostnames, if no hosts
+	// are defined, then set to "*". If the listenerHostname is defined,
+	// then the route must match the Gateway hostname.
+	if len(hostnames) == 0 && listenerHostname == nil {
+		hosts["*"] = struct{}{}
 		return hosts, nil
 	}
 
+	if listenerHostname != nil {
+		if string(*listenerHostname) != "*" {
+
+			// Validate listener hostname.
+			if err := validHostName(string(*listenerHostname)); err != nil {
+				return hosts, []error{err}
+			}
+
+			if len(hostnames) == 0 {
+				hosts[string(*listenerHostname)] = struct{}{}
+				return hosts, nil
+			}
+		}
+	}
+
 	for _, host := range hostnames {
 
 		hostname := string(host)
+
+		//// A "*" hostname matches anything.
+		//if hostname == "*" {
+		//	hosts[hostname] = struct{}{}
+		//	continue
+		//}
+
+		// Validate the hostname.
 		if err := validHostName(hostname); err != nil {
 			errors = append(errors, err)
 			continue
 		}
-		hosts = append(hosts, string(host))
+
+		if listenerHostname != nil {
+			lhn := string(*listenerHostname)
+
+			// A "*" hostname matches anything.
+			if lhn == "*" {
+				hosts[hostname] = struct{}{}
+				continue
+			} else if lhn == hostname {
+				// If the listener.hostname matches then no need to
+				// do any other validation.
+				hosts[hostname] = struct{}{}
+				continue
+			} else if strings.Contains(lhn, "*") {
+
+				if removeFirstDNSLabel(lhn) != removeFirstDNSLabel(hostname) {
+					errors = append(errors, fmt.Errorf("gateway hostname %q does not match route hostname %q", lhn, hostname))
+					continue
+				}
+			} else {
+				// Validate the gateway listener hostname matches the hostnames hostname.
+				errors = append(errors, fmt.Errorf("gateway hostname %q does not match route hostname %q", lhn, hostname))
+				continue
+			}
+		}
+		hosts[hostname] = struct{}{}
 	}
 	return hosts, errors
 }
 
+func removeFirstDNSLabel(input string) string {
+	return input[strings.IndexAny(input, ".")+1:]
```

Comment:
i think the one thing to be careful about with this is this case:

```
package main

import (
	"fmt"
	"strings"
)

func main() {
	input := "*.foo"
	fmt.Println(input[strings.IndexAny(input, ".")+1:])
	input = "foo"
	fmt.Println(input[strings.IndexAny(input, ".")+1:])
}
```

these both end up being `foo`

so to be safe, just return the whole string if it has no `.` and keep the `.` if it does?

Reply:
nope makes sense!